Source/WebKit/UIProcess/Automation/WebAutomationSession.h

```diff
@@ -113,6 +113,28 @@ class WebAutomationSession final : public API::ObjectImpl<API::Object::Type::Aut
     WebAutomationSession();
     ~WebAutomationSession();
 
+#if ENABLE(REMOTE_INSPECTOR)
+// Maintain the behavior from API::ObjectImpl even when this class multiple inherits from RemoteAutomationTarget.
+    void ref() const
+    {
+        API::ObjectImpl<API::Object::Type::AutomationSession>::ref();
+    }
+    void deref() const
+    {
+        API::ObjectImpl<API::Object::Type::AutomationSession>::deref();
+    }
+#if PLATFORM(COCOA) // Equivalent to `#if DELEGATE_REF_COUNTING_TO_COCOA` in API::ObjectImpl.
+    void* operator new(size_t size)
+    {
+        return API::ObjectImpl<API::Object::Type::AutomationSession>::operator new(size);
+    }
+    void* operator new(size_t size, void* value)
+    {
+        return API::ObjectImpl<API::Object::Type::AutomationSession>::operator new(size, value);
+    }
+#endif
+#endif
```

Comment:
Is this still needed after the latest PR to make `RemoteControllableTarget` inherit from `ThreadSafeRefCounted`?

Reply:
Since `RemoteControllableTarget` now has the methods `ref`, `deref`, and `operator new` (all inherited from `ThreadSafeRefCounted`), I think we will need some form of disambiguation here unless we avoid multiple inheritance altogether.

@cdumez @achristensen07 Chris and Alex, do you have any suggestions on how to write this given I've decided to maintain the multiple inheritance for now?

Reply:
Normally a statement like `using API::ObjectImpl<API::Object::Type::AutomationSession>::ref;` is enough to disambiguate, we don't need to provide a new implementation.

That said, I'm not sure it's the right thing to do here. Doing ref counting from several base classes is generally not safe. One base class's ref count might reach 0 and it would end up calling `delete this;`, not knowing that the ref count from the *other* base class is not 0. Why do we think it is safe here?

We may need to use virtual `ref()` / `deref()` functions in the base class instead of having it subclass ThreadSafeRefCounted.

Reply:
Right, it feels iffy to allow having ref counting applied twice to the same class.

> We may need to use virtual ref() / deref() functions in the base class instead of having it subclass ThreadSafeRefCounted.
Oh, I never thought of this. I made virtual `ref` and `deref` methods for `RemoteControllableTarget` and made  `RemoteInspectionTarget` the one that inherits from `ThreadSafeRefCounted` instead. Thanks!